### PR TITLE
Api Types Refactor

### PR DESCRIPTION
**Pull request by Joerger**
Refactors:
- Moved type related logic from `/lib/services` to `/api/types`, including:
  - types, methods.
  - marshal logic for each type.
  - Irregular moves:
    - the 3 `ExternalIdentity` methods and json schema in `/lib/services/identity.go` were moved to `/api/types/user.go`.
    - Moved function `NewNamespace` from `presence.go` to `namespace.go`. 
- Created `/lib/services/types.go`, which holds type aliases fo all relevant types, functions, and variables moved to the new types package. This simplifies the refactoring and allows for backwards compatibility.
  - types with `_`'s like `types.KeepAlive_APP` could not be type aliased, because drone's linter forbids `_`'s in names.
- Moved defaults and constants needed in the api package to `/api/defaults` and `/api/constants`.
- Added missing comments to all of `/lib/services` and `/api/types`.
- Reordered functions to be more uniform between files.

Changes:
- Role interface - ApplyTraits removed as a method on Role, replaced with an exported function in services. 
- Refactor `resetpasswordtoken.go` and `resetpasswordtokensecrets.go`.
  - "New" functions now return a pointer to a struct `ResetPasswordToken(Secrets)V3` as an interface `ResetPasswordToken(Secrets)`
  - Added a couple setter methods to the interfaces for compatibility.
- Removed all unused V1 structs and refactored marshaling logic to reflect this. only `ProvisionTokenV1` remains.
- Marshaler structs are now private, because they should only be used via their interface and the get/set marshaler methods.
- Deprecated `services.NewCertAuthority` in favor of `types.NewCertAuthority` to take advantage of the `CertAuthoritySpecV2` struct.

Future PR's:
- Break remaining small decencies:
  - `/lib` dependencies: `/utils`, `/utils/parse`,`/events`, `/wrappers`, `/jwt`, `/tlsca`, `/session`, `/backend`
- User experience Improvements with client - Redesign the client with new helper functions, such as `LoadTSHConfig` which will load user credentials directly from profile.
- Move teleport-plugins to the api package, and unify plugins with the single api client
- Introduce new versioning compatibility strategy for the api.
- Improve Documentation with a combination of GoDocs and protoc-gen-doc. Might need to improve comments across the code base for these new documentation sources to be useful.

Reference [RFD](https://github.com/gravitational/teleport/pull/4746) (needs update)

### CODE REVIEW

File: api/proto/auth/authproto.proto [hunk 1]
```diff
@@ -361,93 +359,91 @@ message CreateAppSessionRequest {
...
     // Session is the application web session.
-    services.WebSessionV2 Session = 1 [ (gogoproto.jsontag) = "session" ];
+    types.WebSessionV2 Session = 1 [ (gogoproto.jsontag) = "session" ];
 }
 
 // DeleteAppSessionRequest contains the parameters used to remove an application web session.
 message DeleteAppSessionRequest { string SessionID = 1 [ (gogoproto.jsontag) = "session_id" ]; }
 
-// GetKubeServicesRequest are the parameters used to request kubernetes services.
+// GetKubeServicesRequest are the parameters used to request kubernetes types.
```
Comment (Joerger) [inline api/proto/auth/authproto.proto:368]: Change "types." back to "services." in this comment and those below

File: api/types/duration.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+func (d Duration) Value() time.Duration {
+	return time.Duration(d)
+}
+
+// MarshalJSON marshals Duration to string
+func (d Duration) MarshalJSON() ([]byte, error) {
+	return json.Marshal(fmt.Sprintf("%v", d.Duration()))
+}
+
+// UnmarshalJSON marshals Duration to string
```
Comment (a-palchikov) [inline api/types/duration.go:47]:
```suggestion
// UnmarshalJSON marshals this duration value as a string
```
for naming consistency

File: api/types/duration.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		out, err := time.ParseDuration(stringVar)
+		if err != nil {
+			return trace.BadParameter(err.Error())
+		}
+		*d = Duration(out)
+	}
+	return nil
+}
+
+// MaxDuration returns maximum duration that is possible
```
Comment (a-palchikov) [inline api/types/duration.go:92]:
```suggestion
// MaxDuration returns the maximum duration value
```

File: api/types/parser.go [hunk 4]
```diff
@@ -223,45 +223,6 @@ func (ctx *Context) GetIdentifier(fields []string) (interface{}, error) {
 	}
 }
 
-// NewParserFn returns function that creates parser of 'where' section
```
Comment (a-palchikov) [inline api/types/parser.go:226]: Sorry for putting this out of context - can you also fix the typo in RuleContext.GetResource:
unpecified -> unspecified
?

File: api/types/parser.go [hunk 5]
```diff
@@ -340,3 +301,42 @@ func (r *EmptyResource) GetName() string {
...
 }
+
+// NewParserFn returns function that creates parser of 'where' section
+// in access rules
+type NewParserFn func(ctx RuleContext) (predicate.Parser, error)
+
+var whereParser = NewWhereParser
+var actionsParser = NewActionsParser
+
+// GetWhereParserFn returns global function that creates where parsers
```
Comment (a-palchikov) [inline api/types/parser.go:312]:
```suggestion
// GetWhereParserFn returns the global function that constructs WHERE predicate parsers
```

File: api/types/presence.go [hunk 6]
```diff
@@ -200,15 +194,10 @@ type Presence interface {
...
-// NewNamespace returns new namespace
-func NewNamespace(name string) Namespace {
-	return Namespace{
-		Kind:    KindNamespace,
-		Version: V2,
-		Metadata: Metadata{
-			Name: name,
-		},
-	}
+// ProxyGetter is an service that gets proxies
```
Comment (a-palchikov) [inline api/types/presence.go:197]:
```suggestion
// ProxyGetter is a service that gets proxies
```

File: api/types/presence.go [hunk 7]
```diff
@@ -240,6 +229,7 @@ func (s *KeepAlive) GetType() string {
 	}
 }
 
+// CheckAndSetDefaults checks and sets default values
```
Comment (a-palchikov) [inline api/types/presence.go:232]:
```suggestion
// CheckAndSetDefaults validates this KeepAlive value and sets default values
```

File: api/types/provisioning.go [hunk 8]
```diff
@@ -95,33 +66,39 @@ type ProvisionToken interface {
...
 
-// ProvisionTokensFromV1 converts V1 provision tokens to resource list
-func ProvisionTokensFromV1(in []ProvisionTokenV1) []ProvisionToken {
-	if in == nil {
-		return nil
-	}
-	out := make([]ProvisionToken, len(in))
-	for i := range in {
-		out[i] = in[i].V2()
+// MustCreateProvisionToken returns a new valid provision token
```
Comment (a-palchikov) [inline api/types/provisioning.go:89]:
```suggestion
// MustCreateProvisionToken returns a new provision token from the specified value, roles and the given expiration.
// Panics on errors. Only used in tests.
```

File: api/types/provisioning.go [hunk 9]
```diff
@@ -235,6 +212,30 @@ func (p ProvisionTokenV2) String() string {
...
+		return nil
+	}
+	out := make([]ProvisionTokenV1, len(in))
+	for i := range in {
+		out[i] = *in[i].V1()
+	}
+	return out
+}
+
+// ProvisionTokensFromV1 converts V1 provision tokens to resource list
```
Comment (a-palchikov) [inline api/types/provisioning.go:227]:
```suggestion
// ProvisionTokensFromV1 converts V1 provision tokens to the current version of the resource
```

File: api/types/resetpasswordtoken.go [hunk 10]
```diff
@@ -42,6 +44,18 @@ type ResetPasswordToken interface {
 	CheckAndSetDefaults() error
 }
 
+// NewResetPasswordToken is a convenience wa to create a RemoteCluster resource.
```
Comment (a-palchikov) [inline api/types/resetpasswordtoken.go:47]:
```suggestion
// NewResetPasswordToken is a convenience way to create a RemoteCluster resource.
```

File: api/types/resetpasswordtoken.go [hunk 11]
```diff
@@ -172,21 +172,8 @@ type ResetPasswordTokenMarshaler interface {
...
-	resetPasswordTokenMarshaler = m
-}
-
-// GetResetPasswordTokenMarshaler returns ResetPasswordToken marshaler
-func GetResetPasswordTokenMarshaler() ResetPasswordTokenMarshaler {
-	marshalerMutex.Lock()
-	defer marshalerMutex.Unlock()
-	return resetPasswordTokenMarshaler
-}
+// TeleportResetPasswordTokenMarshaler implements ResetPasswordTokenMarshaler
```
Comment (a-palchikov) [inline api/types/resetpasswordtoken.go:175]:
```suggestion
// TeleportResetPasswordTokenMarshaler marshals reset password tokens to/from JSON-encoded payloads.
// Implements ResetPasswordTokenMarshaler.
```

File: api/types/resetpasswordtokensecrets.go [hunk 12]
```diff
@@ -39,6 +41,24 @@ type ResetPasswordTokenSecrets interface {
 	CheckAndSetDefaults() error
 }
 
+// NewResetPasswordTokenSecrets creates an instance of ResetPasswordTokenSecrets
+func NewResetPasswordTokenSecrets(tokenID string) (ResetPasswordTokenSecretsV3, error) {
```
Comment (a-palchikov) [inline api/types/resetpasswordtokensecrets.go:45]: nit: why not return a pointer? Isn't this implementing the `ResetPasswordToken` interface?

File: api/types/resetpasswordtokensecrets.go [hunk 13]
```diff
@@ -192,21 +192,8 @@ type ResetPasswordTokenSecretsMarshaler interface {
...
-	resetPasswordTokenSecretsMarshaler = m
-}
-
-// GetResetPasswordTokenSecretsMarshaler returns ResetPasswordTokenSecrets marshaler
-func GetResetPasswordTokenSecretsMarshaler() ResetPasswordTokenSecretsMarshaler {
-	marshalerMutex.Lock()
-	defer marshalerMutex.Unlock()
-	return resetPasswordTokenSecretsMarshaler
-}
+// TeleportResetPasswordTokenSecretsMarshaler implements ResetPasswordTokenSecretsMarshaler
```
Comment (a-palchikov) [inline api/types/resetpasswordtokensecrets.go:195]:
```suggestion
// TeleportResetPasswordTokenSecretsMarshaler marshals reset password tokens to/from JSON-encoded payloads.
// Iimplements ResetPasswordTokenSecretsMarshaler.
```

File: api/types/role.go [hunk 14]
```diff
@@ -0,0 +1,1204 @@
...
+		}
+	}
+	return nil
+}
+
+// score is a sorting score of the rule, the more the score, the more
+// specific the rule is
+func (r *Rule) score() int {
+	score := 0
+	// wilcard rules are less specific
```
Comment (a-palchikov) [inline api/types/role.go:603]:
```suggestion
	// wildcard rules are less specific
```

Comment (a-palchikov) [inline api/types/role.go:603]:
```suggestion
	// wildcard rules are less specific
```

File: api/types/trustedcluster.go [hunk 15]
```diff
@@ -441,6 +326,140 @@ func (c *TrustedClusterV2) String() string {
...
+		}
+		if len(roleMap.Local) == 0 {
+			return nil, trace.BadParameter("missing 'local' parameter for 'role_map'")
+		}
+		for _, local := range roleMap.Local {
+			if local == "" {
+				return nil, trace.BadParameter("missing 'local' property of 'role_map' entry")
+			}
+			if local == constants.Wildcard {
+				return nil, trace.BadParameter("wilcard value is not supported for 'local' property of 'role_map' entry")
```
Comment (a-palchikov) [inline api/types/trustedcluster.go:387]:
```suggestion
				return nil, trace.BadParameter("wildcard value is not supported for 'local' property of 'role_map' entry")
```

File: api/types/trustedcluster.go [hunk 16]
```diff
@@ -70,8 +71,8 @@ type TrustedCluster interface {
 // NewTrustedCluster is a convenience wa to create a TrustedCluster resource.
```
Comment (a-palchikov) [inline api/types/trustedcluster.go:71]:
```suggestion
// NewTrustedCluster is a convenience way to create a TrustedCluster resource.
```

File: api/types/trustedcluster.go [hunk 17]
```diff
@@ -441,6 +326,140 @@ func (c *TrustedClusterV2) String() string {
...
+}
+
+// Map maps local roles to remote roles
+func (r RoleMap) Map(remoteRoles []string) ([]string, error) {
+	_, err := r.parse()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var outRoles []string
+	// when no remote roles is specified, assume that
```
Comment (a-palchikov) [inline api/types/trustedcluster.go:406]:
```suggestion
	// when no remote roles are specified, assume that
```

File: api/types/tunnel.go [hunk 18]
```diff
@@ -196,15 +185,37 @@ func (r *ReverseTunnelV2) Check() error {
...
-		if err != nil {
+		if _, err := utils.ParseAddr(addr); err != nil {
 			return trace.Wrap(err)
 		}
 	}
 
 	return nil
 }
 
+// TunnelType is the type of tunnel. Either node or proxy.
```
Comment (a-palchikov) [inline api/types/tunnel.go:196]:
```suggestion
// TunnelType is the type of tunnel.
```
I think there're more tunnel types which rendered the comment obsolete.

File: lib/services/role.go [hunk 19]
```diff
@@ -2034,6 +1326,7 @@ func (set RoleSet) String() string {
 	return fmt.Sprintf("roles %v", strings.Join(roleNames, ","))
 }
 
+// CheckAccessToRule checks if the RoleSet provides access
```
Comment (a-palchikov) [inline lib/services/role.go:1329]:
```suggestion
// CheckAccessToRule checks if the RoleSet provides access in the given namespace to the specified resource and verb.
// silent controls whether the access violations are logged.
```

File: lib/services/types.go [hunk 20]
```diff
@@ -0,0 +1,594 @@
...
+// user.go
+type User = types.User
+
+var (
+	NewUser          = types.NewUser
+	SetUserMarshaler = types.SetUserMarshaler
+	GetUserMarshaler = types.GetUserMarshaler
+)
+
+// The following Constants are imported from api/constants to simplify
```
Comment (a-palchikov) [inline lib/services/types.go:528]:
```suggestion
// The following constants are imported from api/constants to simplify
```

File: api/types/role.go [hunk 21]
```diff
@@ -0,0 +1,1204 @@
...
+	defer marshalerMutex.Unlock()
+	roleMarshaler = m
+}
+
+// GetRoleMarshaler returns currently set RoleMarshaler
+func GetRoleMarshaler() RoleMarshaler {
+	marshalerMutex.Lock()
+	defer marshalerMutex.Unlock()
+	return roleMarshaler
+}
```
Comment (russjones) [inline api/types/role.go:1204]: @knisbet @r0mant This was used by Gravity right? Thinking about removing it if not needed by Gravity anymore.

File: lib/services/types.go [hunk 22]
```diff
@@ -0,0 +1,594 @@
+package services
```
Comment (awly) [inline lib/services/types.go:1]: Missing license header

**Review by awly — COMMENTED** [review]
FYI, you can selectively disable linter checks with https://golangci-lint.run/usage/false-positives/#nolint
That would let you make aliases with underscores in `services`.
But it's probably not worth it at this point.

Why is `webassets` submodule updated?

File: api/types/authentication.go [hunk 23]
```diff
@@ -83,10 +84,10 @@ type AuthPreference interface {
 // NewAuthPreference is a convenience method to to create AuthPreferenceV2.
 func NewAuthPreference(spec AuthPreferenceSpecV2) (AuthPreference, error) {
 	return &AuthPreferenceV2{
-		Kind:    KindClusterAuthPreference,
-		Version: V2,
+		Kind:    constants.KindClusterAuthPreference,
+		Version: constants.V2,
```
[the remark this answers is not in the document]
Comment (Joerger) [reply api/types/authentication.go:88]: I think I had a reason a few weeks ago but this is no longer necessary, i'll move it to `types/constants.go`

File: api/types/constants.go [hunk 24]
```diff
@@ -52,7 +51,7 @@ const (
 	// KindRole is a role resource
 	KindRole = "role"
 
-	// KindAccessRequest is an AccessRequest resource
+	// KindAccessRequest is an AccessReqeust resource
```
Comment (a-palchikov) [inline api/types/constants.go:54]: this reverted back to a typo.

File: api/types/oidc.go [hunk 25]
```diff
@@ -648,68 +474,154 @@ func OIDCClaimsToTraits(claims jose.Claims) map[string][]string {
...
+	  }
+	},
+	"claims_to_roles": {
+	  "type": "array",
+	  "items": %v
+	}
+  }
+}`, ClaimMappingSchema)
+
+// OIDCConnectorV2SchemaTemplate is a template JSON Schema for user
```
Comment (a-palchikov) [inline api/types/oidc.go:506]:
```suggestion
// OIDCConnectorV2SchemaTemplate is a template JSON Schema for OIDC connector
```

File: api/types/oidc.go [hunk 26]
```diff
@@ -648,68 +474,154 @@ func OIDCClaimsToTraits(claims jose.Claims) map[string][]string {
...
-	RedirectURL string `json:"redirect_url"`
-	// Display - Friendly name for this provider.
-	Display string `json:"display"`
-	// Scope is additional scopes set by provider
-	Scope []string `json:"scope"`
-	// ClaimsToRoles specifies dynamic mapping from claims to roles
-	ClaimsToRoles []ClaimMapping `json:"claims_to_roles"`
+var connectorMarshaler OIDCConnectorMarshaler = &teleportOIDCConnectorMarshaler{}
+
+// SetOIDCConnectorMarshaler sets global user marshaler
```
Comment (a-palchikov) [inline api/types/oidc.go:538]:
```suggestion
// SetOIDCConnectorMarshaler sets global OIDC connector marshaler
```

File: api/types/oidc.go [hunk 27]
```diff
@@ -648,68 +474,154 @@ func OIDCClaimsToTraits(claims jose.Claims) map[string][]string {
...
+func SetOIDCConnectorMarshaler(m OIDCConnectorMarshaler) {
+	marshalerMutex.Lock()
+	defer marshalerMutex.Unlock()
+	connectorMarshaler = m
 }
 
-// V1 returns V1 version of the resource
-func (o *OIDCConnectorV1) V1() *OIDCConnectorV1 {
-	return o
+// GetOIDCConnectorMarshaler returns currently set user marshaler
```
Comment (a-palchikov) [inline api/types/oidc.go:545]:
```suggestion
// GetOIDCConnectorMarshaler returns currently set OIDC connector marshaler
```

File: api/types/oidc.go [hunk 28]
```diff
@@ -648,68 +474,154 @@ func OIDCClaimsToTraits(claims jose.Claims) map[string][]string {
...
-		Spec: OIDCConnectorSpecV2{
-			IssuerURL:     o.IssuerURL,
-			ClientID:      o.ClientID,
-			ClientSecret:  o.ClientSecret,
-			RedirectURL:   o.RedirectURL,
-			Display:       o.Display,
-			Scope:         o.Scope,
-			ClaimsToRoles: o.ClaimsToRoles,
-		},
+// OIDCConnectorMarshaler implements marshal/unmarshal of User implementations
```
Comment (a-palchikov) [inline api/types/oidc.go:552]:
```suggestion
// OIDCConnectorMarshaler implements marshal/unmarshal of OIDCConnector implementations
```

File: api/types/resource.go [hunk 29]
```diff
@@ -0,0 +1,518 @@
...
+	}
+	m, err := marshal(resource, opts...)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return m, nil
+}
+
+// UnmarshalResource attempts to unmarshal a resource dynamically, returning NotImplementedError
+// if not unmarshaler has been registered.
```
Comment (a-palchikov) [inline api/types/resource.go:444]:
```suggestion
// if no unmarshaler has been registered.
```

File: api/types/role.go [hunk 30]
```diff
@@ -0,0 +1,1203 @@
...
+			_, err = parser.Parse(action)
+			if err != nil {
+				return trace.BadParameter("could not parse action %v %q, error: %v", i, action, err)
+			}
+		}
+	}
+	return nil
+}
+
+// score is a sorting score of the rule, the more the score, the more
```
Comment (a-palchikov) [inline api/types/role.go:598]:
```suggestion
// score is a sorting score of the rule, the larger the score, the more
```

**Review by a-palchikov — APPROVED** [review]

**Review by awly — APPROVED** [review]

File: api/defaults/defaults.go [hunk 31]
```diff
@@ -0,0 +1,65 @@
...
+	// set to help keep connections alive when using AWS NLBs (which have a default
+	// timeout of 350 seconds)
+	KeepAliveInterval = 5 * time.Minute
+)
+
+// EnhancedEvents returns the default list of enhanced events.
+func EnhancedEvents() []string {
+	return []string{
+		teleport.EnhancedRecordingCommand,
+		teleport.EnhancedRecordingNetwork,
```
Comment (russjones) [inline api/defaults/defaults.go:63]: You're already moving constants over, I'd move these over as well and break the dependency on `github.com/gravitational/teleport`.

File: examples/go-client/access_workflow.go [hunk 32]
```diff
@@ -44,7 +45,7 @@ func accessWorkflow(ctx context.Context, client *auth.Client) {
 	}()
 
 	// retrieve all pending access requests
-	filter := services.AccessRequestFilter{State: services.RequestState_PENDING}
+	filter := services.AccessRequestFilter{State: types.RequestState_PENDING}
```
Comment (russjones) [inline examples/go-client/access_workflow.go:48]: Do the docs pull from this file? If so I'd leave it `services` for now until you're fully done. If someone deploys from master, the docs will be wrong.

**Review by russjones — APPROVED** [review]
To minor things then you're good to merge.

File: api/defaults/defaults.go (hunk 31 above)
Comment (Joerger) [reply api/defaults/defaults.go:63]: `github.com/gravitational/teleport` is still being imported in several other spots in the api package, so I think at this point it'd be better served as a separate pr.

Comment (russjones) [reply api/defaults/defaults.go:63]: Okay sounds good to me.

File: api/types/oidc.go [hunk 33]
```diff
@@ -109,124 +109,6 @@ func NewOIDCConnector(name string, spec OIDCConnectorSpecV2) OIDCConnector {
...
-		return nil, trace.Wrap(err)
-	}
-	switch h.Version {
-	case "":
-		var c OIDCConnectorV1
-		err := json.Unmarshal(bytes, &c)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		return c.V2(), nil
```
[the remark this answers is not in the document]
Comment (webvictim) [reply api/types/oidc.go:162]: @benarent

**Pull request merged**